src/ArrayTemplates.hs

```diff
@@ -611,15 +611,15 @@ strTy typeEnv env (StructTy _ [innerType]) =
   ]
 strTy _ _ _ = []
 
-takeAddressOrNot :: TypeEnv -> Ty -> String
-takeAddressOrNot typeEnv t = if isManaged typeEnv t then "&" else ""
+takeAddressOrNot :: TypeEnv -> Env -> Ty -> String
```

Comment:
This is actually kind of wrong, should check if it's a primitive type instead?

Reply:
Couldn't user-defined types also be primitive (like integral types etc)?

Reply:
Yes... we need a special interface for this probably.

Reply:
In this case it's actually just a matter of figuring out whether the str function takes a ref or not, so perhaps the code should just query for both and see if any fits.

Reply:
> Yes... we need a special interface for this probably.

This would be similar to the [Copy trait](https://doc.rust-lang.org/std/marker/trait.Copy.html), no?

Reply:
Yup!

Reply:
Except the reason things are not passed as ref to str/prn is because they are so small. Copy trait/interface could work on any big structure as long as it's memcopyable, I think.

Reply:
This is where `derive` helps I guess, have primitives be `Copy` and if you want it on your structure you can `derive` it.

Reply:
@TimDeve Yeah, let's aim for that! If you want to write a small proposal at some point, that'd be great.

Reply:
Added a small issue for this.